### ISSUE TITLE
Open Redirect Vulnerability

### ISSUE THREAD

**Issue by jnbn** [artesaos/seotools#200]
Unvalidated redirects could cause the web application to redirect the request to a malicious URL. By modifying untrusted URL input to a malicious site, an attacker
may successfully launch a phishing scam, steal user credentials and redirect to malicious sites.

Open redirect vulnerabilities also used for bypassing spam gateways and sending malicious files, links to companies by using your company site.

In this test case, open redirect vulnerability is caused by putting untrusted users inside <meta> tags. 

An attacker can exploit this functionality to inject HTML parameters inside the meta tag.

Current code causes the open redirection vulnerability. It strips HTML tags before putting untrusted input inside meta tags, but it does not check for malicious meta-tag parameters that can be injected by an attacker.

```
private function makeTag($key, $value)
    {
        return '<meta name="'.$this->prefix.strip_tags($key).'" content="'.strip_tags($value).'" />';
    }
```

References:
https://www.trustwave.com/en-us/resources/blogs/spiderlabs-blog/understanding-and-discovering-open-redirect-vulnerabilities/
https://blog.detectify.com/2019/05/16/the-real-impact-of-an-open-redirect/
https://www.netsparker.com/blog/web-security/open-redirection-vulnerability-information-prevention/

I'll be sending a PR for this by working on Opengraph.php and TwitterCards.php but if you have any better suggestions for implementations all are welcome.

**Comment by vinicius73** [comment]
Hi @jnbn 
Thanks for request

I don't understand that vulnerability.

Is a package vulnerability or application vulnerability?

Can you provide an example of the exploration of this?

**Comment by MadMikeyB** [comment]
<link> tags are also vulnerable.

VULNERABLE URL:  `https://www.example.com/search?q=test%22%3e%3cscript%3ealert(1)%3c%2fscript%3eqi9qq`

OUTPUT HTML:
```
<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width
...[SNIP]...
<link rel="canonical" href="https://www.example.com/test"><script>alert(1)</script>qi9qq"/>
```

If in config/seotools.php the following is assumed:

```
        'defaults'       => [
            'title'        => 'Example.com', // set false to total remove
            'description'  => false, // set false to total remove
            'separator'    => ' - ',
            'keywords'     => [],
            'canonical'    => null, // Set null for using Url::current(), set false to total remove
        ],
```

This happens because `Url::current()` is used. `Url::current()` doesn't strip out any sort of script tag, etc, because that isn't it's job.

Now imagine the following attack vector:

User creates user generated content with the following

`<img src="https://www.example.com/search?q=test%22%3e%3cscript%3ewindow.location.href='https://badsite.com'%3c%2fscript%3eqi9qq">`

Even on sites which allow HTML but strip HTML using Purifier, "src" is considered safe on images using even the most strict HTMLPurifier configs, and so this gets through, and anyone who loads the page executes this code.

**Comment by jnbn** [comment]
Unfortunately it's a package vulnerability because if we can't get any precautions here, all users of the package must filter and check their titles etc to avoid any redirection attempts by users who want to manipulate. The same rule should be applied to any meta tags including title and metas to avoid this. 

You can see that it's commonly used by hackers to redirect anyone to their target from a different point of source.

**Comment by jnbn** [comment]
I'd be happy if you can merge https://github.com/artesaos/seotools/pull/201 ASAP

**Comment by vinicius73** [comment]
Closed by #201 thanks a lot @jnbn and @MadMikeyB
